### ISSUE TITLE
Simple Example with CIFAR-10. ImageDataGenerator + 'sparse_categorical_crossentropy'

### ISSUE THREAD

**Issue by airplane2230** [tensorflow/tensorflow#47394]
**System information**
- TensorFlow version (use command below): 2.4

I try example CIFAR-10 with ImageDataGenerator. But it doesn't works. 

My custom CNN Model is not training when I use **ImageDataGenerator + sparse_categorical_crossentropy**.
However, when I use **ImageDataGenerator + categorical_crossentropy**, it works well.

what happen in this code?

Thank you and my code is below.

------------------------------------------

```
from tensorflow.keras.datasets import cifar10
import numpy as np

(x_train, y_train), (x_test, y_test) = cifar10.load_data()

x_mean = np.mean(x_train, axis = (0, 1, 2))
x_std = np.std(x_train, axis = (0, 1, 2))

x_train = (x_train - x_mean) / x_std
x_test = (x_test - x_mean) / x_std

from sklearn.model_selection import train_test_split

x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, 
                                                  test_size = 0.3, random_state = 777)

print('data ready~')

from tensorflow.keras.preprocessing.image import ImageDataGenerator

train_datagen = ImageDataGenerator(horizontal_flip = True,
                                   zoom_range = 0.2,
                                   width_shift_range = 0.1,
                                   height_shift_range = 0.1,
                                   rotation_range = 30,
                                   fill_mode = 'nearest'
                                  )
val_datagen = ImageDataGenerator()

batch_size = 32

train_generator = train_datagen.flow(x_train, y_train,
                                    batch_size = batch_size)
val_generator = val_datagen.flow(x_val, y_val,
                                batch_size = batch_size)


from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPool2D, Dense, Flatten, Activation, BatchNormalization
from tensorflow.keras.optimizers import Adam

model = Sequential()

model.add(Conv2D(filters = 32, kernel_size = 3, padding = 'same', input_shape = (32, 32, 3)))
model.add(BatchNormalization())
model.add(Activation('relu'))
model.add(Conv2D(filters = 32, kernel_size = 3, padding = 'same'))
model.add(BatchNormalization())
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size = (2, 2), strides = 2, padding = 'same'))

model.add(Conv2D(filters = 64, kernel_size = 3, padding = 'same'))
model.add(BatchNormalization())
model.add(Activation('relu'))
model.add(Conv2D(filters = 64, kernel_size = 3, padding = 'same'))
model.add(BatchNormalization())
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size = (2, 2), strides = 2, padding = 'same'))

model.add(Conv2D(filters = 128, kernel_size = 3, padding = 'same'))
model.add(BatchNormalization())
model.add(Activation('relu'))
model.add(Conv2D(filters = 128, kernel_size = 3, padding = 'same'))
model.add(BatchNormalization())
model.add(Activation('relu'))
model.add(MaxPool2D(pool_size = (2, 2), strides = 2, padding = 'same'))

model.add(Flatten())
model.add(Dense(256))
model.add(Activation('relu'))
model.add(Dense(10, activation = 'softmax'))

model.compile(optimizer = Adam(1e-4),
             loss = 'sparse_categorical_crossentropy',
             metrics = ['acc'])

def get_step(train_len, batch_size):
    if(train_len % batch_size > 0):
        return train_len // batch_size + 1
    else:
        return train_len // batch_size

history = model.fit(train_generator,
                    epochs = 100,
                    steps_per_epoch = get_step(len(x_train), batch_size),
                    validation_data = val_generator,
                    validation_steps = get_step(len(x_val), batch_size))
```

**Comment by amogh7joshi** [comment]
To begin with, when replicating this code in a Google Colab runtime, it seemed to function properly, so the bug is likely with your system or for a different reason. The code does not seem to have any issues.

In general, `sparse_categorical_crossentropy` should be used when prediction classes are mutually exclusive, i.e. each training example belongs to one single class, and `categorical_crossentropy` should be used when one sample may have multiple labels or you need the *probability* for each output.

For example, if you have a label of `2`, from the range 0-3, then you should use `sparse_categorical_crossentropy`. However, if your label vector is `[0.1, 0.05, 0.75, 0.1]`, then you should use `categorical_crossentropy` since you are finding the prediction *confidence* for each class.

For the Cifar-10 example that you've given, the labels are each contained in a 1-dimensional label vector, i.e. `[[0], [1], [2], [5]]` (with a shape of (35000, 1). This means that it will not throw an error with `sparse_categorical_crossentropy`, but it would work best with `categorical_crossentropy`. If you want to use `sparse_categorical_crossentropy`, then call either `np.squeeze` or `tf.squeeze` on the data to get rid of the last dimension.

**Comment by airplane2230** [comment]
Firstly, Thank you for your answer And I understand it.

Additional, I check that it works well using np.squeeze on y_train.

But, I don't understand below.

+ x_train = same
+ y_train = (35000, 1)
+ compile loss = sparse_categorical_crossentropy

In model.fit(), 
A.
model.fit(x_train, y_train, ...) --> good

B.
model.fit(train_generator, ...) --> Bad

what is difference between A and B example?
Is the difference due to the way y_train is handled inside the fit function?

**Comment by amogh7joshi** [comment]
Could you please elaborate on the issues you are facing with the `ImageDataGenerator`? I rarely encounter any issues when working with it, and it in fact is supposed to *help* with training, since it performs spatial augmentation on the images. This allows it to generate more diverse data. 

The clearcut difference between the way X_train and y_train are handled inside of the training function is essentially due to the fact that alongside the spatial augmentation, the `ImageDataGenerator` creates image batches beforehand.

Furthermore, I seem to encounter no issues using either of them when testing in Google Colab, so I'm not sure what the problem is.

**Comment by airplane2230** [comment]
Definitely, No error in code.

when using ImageDatagenerator, I got about 10% accuracy.
But, I use y_train on fit function directly, I got about 80% accuracy.
The problem is that performance is not good.

Literally, I think No difference between the two.

Because,
With ImageDataGenerator gives us the shape of y_train like below.

```
test = train_generator.next()
test[1].shape # output's shape: (32, 1)

And, With just using y_train,
y_train.shape # output's shape: (35000, 1)
But in fit function, I think that the shape of y_train turns into (32, 1)
```

However, when using ImageDataGenerator, performance is good with suqeeze, and when y_train is used directly, performance is good without squeeze.

I don't understand this... why?

**Comment by amogh7joshi** [comment]
The explanation for the different shapes, as I mentioned previously, is because the `ImageDataGenerator` converts the data into **image batches**. This means that the data will be converted into smaller sets of the same data, which will then be passed through the model in sequence.

As for your description of accuracies between the `ImageDataGenerator` and the `X_train`/`y_train`, I can't seem to replicate that; I get similar accuracies with both.

I suggest you recheck your code and if your questions have been answered, please be sure to close the issue.

**Comment by airplane2230** [comment]
Thank you and Rechecking Now.

Finally, I close this issue.

**Comment by YorkSu** [comment]
I ran into the same issue. 
System information: Win 10 21h1, python 3.9, TensorFlow 2.6.0
When I use CIFAR10 + ImageDataGenerator + sparse_categorical_crossentropy in TF2.6, I got 10% acc. 
Using MNIST + ImageDataGenerator + sparse_categorical_crossentropy, 99% acc, it works well.
Using Fruits114 + ImageDataGenerator + sparse_categorical_crossentropy, 99% acc, it works well too.
Using CIFAR10 without ImageDataGenerator, it works well too.
Using CIFAR10 + ImageDataGenerator + sparse_categorical_crossentropy in TF2.1, I got 90% acc, it works well too.
This is very strange to me. I am guessing whether some changes made by ImageDataGenerator caused this strange phenomenon.
